File: frontend/src/lib/SentenceBuffer.ts
/**
 * SentenceBuffer — accumulates LLM token deltas and emits complete sentences
 * for the streaming TTS pipeline.
 *
 * Design ref: REALTIME_VOICE_PIPELINE.md §5
 */

export type SentenceCallback = (sentence: string) => void;

export interface SentenceBufferOptions {
  /** Minimum chars before a boundary triggers emission (default 10) */
  minLength?: number;
  /** Force-split at this length if no boundary found (default 300) */
  maxLength?: number;
  /** Wait this long before flushing short sentences (default 500ms) */
  shortTimeoutMs?: number;
}

// Sentence-ending punctuation + optional trailing quotes/parens
const BOUNDARY_RE = /([.!?。！？\n]["'」）)\s]*)/;

// Common abbreviations that end with "." but aren't sentence ends
const ABBREVIATIONS = new Set([
  'Mr', 'Mrs', 'Ms', 'Dr', 'Prof', 'Sr', 'Jr', 'St', 'Ave', 'Blvd',
  'vs', 'etc', 'i.e', 'e.g', 'a.m', 'p.m',
]);

export class SentenceBuffer {
  private buffer = '';
  private onSentence: SentenceCallback;
  private shortTimer: ReturnType<typeof setTimeout> | null = null;
  private streamComplete = false;

  private readonly minLength: number;
  private readonly maxLength: number;
  private readonly shortTimeoutMs: number;

  constructor(onSentence: SentenceCallback, options: SentenceBufferOptions = {}) {
    this.onSentence = onSentence;
    this.minLength = options.minLength ?? 10;
    this.maxLength = options.maxLength ?? 300;
    this.shortTimeoutMs = options.shortTimeoutMs ?? 500;
  }

  /** Add a token delta from the LLM stream */
  addToken(token: string): void {
    this.buffer += token;
    this.clearShortTimer();
    this.tryFlush();
  }

  /** Mark the LLM stream as complete — flush everything remaining */
  markStreamComplete(): void {
    this.streamComplete = true;
    this.clearShortTimer();
    // Try boundary-based flush first (streamComplete bypasses minLength)
    this.tryFlush();
    // Then flush any remaining text that has no boundary
    this.flush();
  }

  /** Cancel everything — used on barge-in or deactivation */
  cancel(): void {
    this.clearShortTimer();
    this.buffer = '';
    this.streamComplete = false;
  }

  /** Force-flush whatever is in the buffer */
  flush(): void {
    this.clearShortTimer();
    const remaining = this.buffer.trim();
    if (remaining.length > 0) {
      this.onSentence(remaining);
    }
    this.buffer = '';
  }

  /** Current buffer contents (for debugging/display) */
  get pending(): string {
    return this.buffer;
  }

  // ── Internal ──────────────────────────────────────────────────────

  private tryFlush(): void {
    // Force-split very long text without boundaries
    if (this.buffer.length > this.maxLength) {
      this.forceSplit();
      return;
    }

    // Scan for the first real sentence boundary (skip abbreviations)
    let searchFrom = 0;
    while (searchFrom < this.buffer.length) {
      const sub = this.buffer.slice(searchFrom);
      const match = sub.match(BOUNDARY_RE);
      if (!match || match.index === undefined) return; // No boundary found

      const boundaryEnd = searchFrom + match.index + match[0].length;
      const sentence = this.buffer.slice(0, boundaryEnd).trim();

      // Check for false boundary (abbreviation like "Dr." "Mr.")
      if (this.isAbbreviation(sentence)) {
        // Skip past this boundary and keep looking
        searchFrom = boundaryEnd;
        continue;
      }

      const rest = this.buffer.slice(boundaryEnd);

      // Check minimum length
      if (sentence.length >= this.minLength || this.streamComplete) {
        this.buffer = rest;
        this.onSentence(sentence);
        // Recursively check if rest has more boundaries
        if (this.buffer.length > 0) {
          this.tryFlush();
        }
      } else {
        // Short sentence — wait for more tokens or timeout
        this.startShortTimer();
      }
      return;
    }
  }

  private forceSplit(): void {
    const splitAt = this.findSplitPoint(this.buffer, this.maxLength);
    const chunk = this.buffer.slice(0, splitAt).trim();
    this.buffer = this.buffer.slice(splitAt).trimStart();
    if (chunk) {
      this.onSentence(chunk);
    }
    // Continue checking remainder
    if (this.buffer.length > this.maxLength) {
      this.forceSplit();
    }
  }

  /** Find a natural split point near maxLen (comma > space > hard cut) */
  private findSplitPoint(text: string, maxLen: number): number {
    const halfMax = Math.floor(maxLen * 0.5);

    // Prefer comma
    const commaIdx = text.lastIndexOf(',', maxLen);
    if (commaIdx > halfMax) return commaIdx + 1;

    // Prefer semicolon / colon
    const semiIdx = text.lastIndexOf(';', maxLen);
    if (semiIdx > halfMax) return semiIdx + 1;
    const colonIdx = text.lastIndexOf(':', maxLen);
    if (colonIdx > halfMax) return colonIdx + 1;

    // Fall back to space
    const spaceIdx = text.lastIndexOf(' ', maxLen);
    if (spaceIdx > halfMax) return spaceIdx + 1;

    // Hard cut
    return maxLen;
  }

  /** Check if the "." at the end is an abbreviation, not a sentence end */
  private isAbbreviation(sentence: string): boolean {
    if (!sentence.endsWith('.')) return false;
    // Extract the last word before the period
    const words = sentence.trimEnd().split(/\s+/);
    const lastWord = words[words.length - 1]?.replace(/\.$/, '');
    if (!lastWord) return false;
    return ABBREVIATIONS.has(lastWord);
  }

  private startShortTimer(): void {
    if (this.shortTimer) return; // Already waiting
    this.shortTimer = setTimeout(() => {
      this.shortTimer = null;
      const pending = this.buffer.trim();
      if (pending) {
        this.buffer = '';
        this.onSentence(pending);
      }
    }, this.shortTimeoutMs);
  }

  private clearShortTimer(): void {
    if (this.shortTimer) {
      clearTimeout(this.shortTimer);
      this.shortTimer = null;
    }
  }
}

File: services/chat-service/app/services/sentence_buffer.py
"""SentenceBuffer — accumulates LLM tokens into complete sentences for TTS.

Synchronous, server-side version. No timers — the caller drives the pipeline
by pushing tokens and collecting emitted sentences.

Design ref: VOICE_PIPELINE_V2.md §3.7
"""
from __future__ import annotations

import re


# Sentence-ending punctuation followed by optional quotes/parens and whitespace
_SENTENCE_END = re.compile(r'([.!?]["\')\]]*)\s')
_CJK_SENTENCE_END = re.compile(r'([。！？]["\'」）]*)\s?')

# Common abbreviations that end with "." but are NOT sentence boundaries
_ABBREVIATIONS = frozenset([
    'Mr', 'Mrs', 'Ms', 'Dr', 'Prof', 'Sr', 'Jr', 'St', 'Ave', 'Blvd',
    'vs', 'etc', 'i.e', 'e.g', 'a.m', 'p.m',
])

# Clause delimiters for voice mode (lower latency splitting)
_CLAUSE_DELIMS = [', ', ' — ', '; ', ' but ', ' and ', ' so ', ' because ']
_CJK_CLAUSE_DELIMS = ['、', '，']

_CLAUSE_MIN_LENGTH = 40
_MAX_LENGTH = 300


class SentenceBuffer:
    """Accumulate LLM tokens, emit complete sentences for TTS."""

    def __init__(self, clause_mode: bool = False):
        self._buffer = ''
        self._clause_mode = clause_mode

    @property
    def pending(self) -> str:
        return self._buffer

    def push(self, token: str) -> list[str]:
        """Push a token, return list of complete sentences (may be empty)."""
        self._buffer += token
        sentences: list[str] = []

        while True:
            # Force-split very long text without boundaries
            if len(self._buffer) > _MAX_LENGTH:
                split_at = self._find_split_point(self._buffer, _MAX_LENGTH)
                chunk = self._buffer[:split_at].strip()
                self._buffer = self._buffer[split_at:].lstrip()
                if chunk:
                    sentences.append(chunk)
                continue

            # Try sentence boundaries
            emitted = self._try_sentence_boundary()
            if emitted:
                sentences.append(emitted)
                continue

            # Try clause-mode splits (voice only, 40+ chars)
            if self._clause_mode:
                emitted = self._try_clause_split()
                if emitted:
                    sentences.append(emitted)
                    continue

            break

        return sentences

    def flush(self) -> str | None:
        """Flush remaining buffer (call when LLM stream ends).

        Returns the remaining text, or None if empty.
        """
        remaining = self._buffer.strip()
        self._buffer = ''
        return remaining if remaining else None

    def _try_sentence_boundary(self) -> str | None:
        """Try to find and emit at a sentence boundary."""
        search_from = 0
        while search_from < len(self._buffer):
            # Try standard sentence ends
            match = _SENTENCE_END.search(self._buffer, search_from)
            if not match:
                # Try CJK sentence ends
                match = _CJK_SENTENCE_END.search(self._buffer, search_from)
            if not match:
                return None

            end = match.end()
            sentence = self._buffer[:end].strip()

            # Check abbreviation
            if self._is_abbreviation(sentence):
                search_from = end
                continue

            self._buffer = self._buffer[end:]
            return sentence

        return None

    def _try_clause_split(self) -> str | None:
        """Try clause-level splitting when buffer is long enough."""
        if len(self._buffer) <= _CLAUSE_MIN_LENGTH:
            return None

        all_delims = _CLAUSE_DELIMS + _CJK_CLAUSE_DELIMS
        for delim in all_delims:
            idx = self._buffer.rfind(delim)
            if idx > _CLAUSE_MIN_LENGTH:
                end = idx + len(delim)
                clause = self._buffer[:end].strip()
                self._buffer = self._buffer[end:]
                return clause

        return None

    @staticmethod
    def _find_split_point(text: str, max_len: int) -> int:
        """Find a natural split point near max_len."""
        half = max_len // 2
        for char in [',', ';', ':', ' ']:
            idx = text.rfind(char, half, max_len)
            if idx > half:
                return idx + 1
        return max_len

    @staticmethod
    def _is_abbreviation(sentence: str) -> bool:
        """Check if the trailing '.' is an abbreviation, not a sentence end."""
        trimmed = sentence.rstrip()
        if not trimmed.endswith('.'):
            return False
        words = trimmed.split()
        if not words:
            return False
        last_word = words[-1].rstrip('.')
        return last_word in _ABBREVIATIONS
